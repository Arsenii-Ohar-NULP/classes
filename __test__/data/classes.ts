import Class from 'components/classes/Class';

export const sampleFiveClasses: Class[] = [
    {
        id: 1,
        title: 'Class 1',
        description: 'This is class 1',
        teacher_id: 1,
        students_number: 1
    },
    {
        id: 2,
        title: 'Class 2',
        description: 'This is class 2',
        teacher_id: 2,
        students_number: 1
    },
    {
        id: 3,
        title: 'Class 3',
        description: 'This is class 3',
        teacher_id: 3,
        students_number: 1

    },
    {
        id: 4,
        title: 'Class 4',
        description: 'This is class 4',
        teacher_id: 4,
        students_number: 1
    },
    {
        id: 5,
        title: 'Class 5',
        description: 'This is class 5',
        teacher_id: 5,
        students_number: 1
    }
]