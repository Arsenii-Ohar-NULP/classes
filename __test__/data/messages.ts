import Message from 'pages/class/Message';

const sampleMessages : Message[] = [
    {
        id: 1,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 2,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 3,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    }

]