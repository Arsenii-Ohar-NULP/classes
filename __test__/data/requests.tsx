import { JoinRequest } from 'pages/class/JoinRequest'

export const getRequestsWithClassId = (classId: number, number: number) => {
    const requests: JoinRequest[] = []
    for (let i = 0; i < number; i++){
        requests.push(
            {
                userId: i + 1,
                classId: classId
            }
        )
    }

    return requests;
}