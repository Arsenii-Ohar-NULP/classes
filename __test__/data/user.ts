import User, { Role } from 'components/account/User';

export const sampleUser: User = {
    id: 1,
    username: 'seniorohar',
    firstName: 'Arsenii',
    lastName: 'Ohar',
    email: 'arsen.ogar@gmail.com',
    phone: '+380987669293',
    role: Role.Teacher   
}

export const sampleFiveUsers: User[] = [
    {
        id: 1,
        username: 'seniorohar',
        firstName: 'Arsenii',
        lastName: 'Ohar',
        email: 'arsen.ogar@gmail.com',
        phone: '+380987669293',
        role: Role.Student
    },
    {
        id: 2,
    username: 'vitalikpikh',
    firstName: 'Vitalik',
    lastName: 'Pikhotskiy',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
    {
        id: 3,
    username: 'svyatoslavshainoha',
    firstName: 'Svyatoslav',
    lastName: 'Shainoha',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
      {
        id: 4,
    username: 'olehrysin',
    firstName: 'Oleh',
    lastName: 'Rysin',
    email: 'olehrysin@gmail.com',
    phone: '+380967323273',
    role: Role.Student
    },
    {
        id: 5,
    username: 'johnadams',
    firstName: 'John',
    lastName: 'Adams',
    email: 'johnadams@gmail.com',
    phone: '+380967723273',
    role: Role.Student
    }
]