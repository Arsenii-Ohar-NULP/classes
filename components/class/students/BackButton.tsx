import React from 'react';
import {useRouter} from "next/router";
import {Button} from "react-bootstrap";
import backIcon from 'icons/back.svg';
import Image from "next/image";

export const BackButton = () => {
    const router = useRouter();

    const goBack = async () => {
        router.back();
    }

    return <Button variant={'outline-light'} className={'p-1 rounded-pill'} onClick={goBack}>
        <Image src={backIcon} alt={'Back'} height={24} width={24} />
    </Button>
}