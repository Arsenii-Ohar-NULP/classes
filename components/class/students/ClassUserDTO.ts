export interface ClassUserDTO{
        class: number;
        student: number;
}

