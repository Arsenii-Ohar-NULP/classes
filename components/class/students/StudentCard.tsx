import React, {useState} from 'react';
import User from "../../account/User";
import clsx from "clsx";
import ProfilePicture from "../../ProfilePicture";
import styles from 'components/class/students/students.module.scss';
import {DeleteStudentButton} from "./DeleteStudentButton";
import {VisibleOnHover} from "./VisibleOnHover";

interface StudentCardProps {
    student: User;
    onDelete: VoidFunction
}

export const StudentCard = ({student, onDelete}: StudentCardProps) => {
    const [isHover, setIsHover] = useState<boolean>();
    return <tr
        className={clsx('p-2 w-100 text-start overflow-scroll', styles['neat-text'], styles['student'])}
        onMouseOver={() => setIsHover(true)}
        onMouseOut={() => setIsHover(false)}>
        <td scope={'row'} className={'d-flex text-center gap-2 align-items-center'}>
            <ProfilePicture user={student} hoverOn={false}/>
            <div>{student.firstName} {student.lastName}</div>
        </td>
        <td scope={'row'}>{student.email}</td>
        <td scope={'row'}>
            <div className={'d-flex gap-3'}>
                {student.phone}
                <VisibleOnHover isHover={isHover}>
                    <DeleteStudentButton onDelete={onDelete}/>
                </VisibleOnHover>
            </div>
        </td>
    </tr>
}