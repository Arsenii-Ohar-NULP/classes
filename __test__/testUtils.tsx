import React, { PropsWithChildren } from 'react';
import { render, renderHook } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import type { RootState, AppStore } from 'pages/redux/store';
// As a basic setup, import your same slice reducers
import { authInitialState, authSlice } from 'pages/redux/auth';
import { classSlice, classesInitialState } from 'pages/redux/classes';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderHookWithProviders(
  ui,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...renderHook(() => ui(), { wrapper: Wrapper, ...renderOptions }) };
}
