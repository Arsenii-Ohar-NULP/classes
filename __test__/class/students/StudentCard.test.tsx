import React from 'react';
import {renderWithProviders} from "../../testUtils";
import {sampleUser} from "../../data/user";
import {StudentCard} from "../../../components/class/students/StudentCard";
import {fireEvent, screen, waitFor} from "@testing-library/react";
import {act} from "react-dom/test-utils";

const pushMock = jest.fn();
jest.mock('next/router', () => ({
  useRouter: () => ({
    pathname: '/',
    push: pushMock,
  }),
}));
describe('Student Card', () => {
    it('should match a snapshot', () => {
        const user = sampleUser;
        const onDelete = jest.fn();
        const card = renderWithProviders(<StudentCard student={user} onDelete={onDelete}/>);

        expect(card).toMatchSnapshot();
    })

    it('should output user info', () => {
        const user = sampleUser;
        const onDelete = jest.fn();
        renderWithProviders(<StudentCard student={user} onDelete={onDelete}/>);
        screen.findByText(`${user.firstName} ${user.lastName}`);
        screen.findByText(user.email);
        screen.findByText(user.phone);
    })

    it('should show delete button, when the mouse is over card', async () => {
        const user = sampleUser;
        const onDelete = jest.fn();
        renderWithProviders(<StudentCard student={user} onDelete={onDelete}/>);
        fireEvent.mouseOver(screen.getByRole('row'));

        await waitFor(() => {
            const button = screen.getByTestId('invisible-button');
            expect(button).toHaveStyle('visibility: visible');
        })
    })

    it('should have invisible delete button, when mouse is not over the card', async () => {
         const user = sampleUser;
        const onDelete = jest.fn();
        renderWithProviders(<StudentCard student={user} onDelete={onDelete}/>);
        await act(async () => {
        fireEvent.mouseOut(await screen.findByTestId('student-card'));
        })
        await waitFor(() => {
            const button = screen.getByTestId('invisible-button');
            expect(button).toHaveClass('invisible');
        })
    })

    it('should call onDelete, when delete button is pressed', async () => {
        const user = sampleUser;
        const onDelete = jest.fn();
        renderWithProviders(<StudentCard student={user} onDelete={onDelete}/>)

        await act(async() => {
            fireEvent.mouseOver(await screen.findByTestId('student-card'));
            fireEvent.click(await screen.findByRole('button'));
        })

        await waitFor(() => {
            expect(onDelete).toHaveBeenCalled();
        })
    })


})