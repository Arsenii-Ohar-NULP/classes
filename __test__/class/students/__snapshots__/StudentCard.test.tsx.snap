// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Card should match a snapshot 1`] = `
<div>
  <tr
    class="p-2 w-100 text-start overflow-scroll neat-text student"
    data-testid="student-card"
  >
    <td
      class="d-flex text-center gap-2 align-items-center"
      scope="row"
    >
      <img
        alt="Profile picture for seniorohar"
        class="rounded-circle bg-light border border-primary mx-2 "
        height="48"
        id="profilePic"
        src="https://api.dicebear.com/6.x/lorelei/svg/seed=seniorohar"
        width="48"
      />
      <div>
        Arsenii
         
        Ohar
      </div>
    </td>
    <td
      scope="row"
    >
      arsen.ogar@gmail.com
    </td>
    <td
      scope="row"
    >
      <div
        class="d-flex gap-3"
      >
        +380987669293
        <div
          class="invisible"
          data-testid="invisible-button"
        >
          <button
            class="p-1 btn btn-danger"
            type="button"
          >
            <img
              alt="Delete Icon"
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
              style="color: transparent;"
              width="24"
            />
          </button>
        </div>
      </div>
    </td>
  </tr>
</div>
`;
